### ISSUE TITLE
PCA memory requirements

### ISSUE THREAD

**Issue by concatenize** [morris-lab/CellOracle#17]
I'm trying CellOracle using an scRNA dataset with ~50k cells. Running PCA is taking over an hour, and currently (it's still running) uses 34G of memory. Is this expected? With SCANPY, this would be much quicker, right?

**Comment by concatenize** [comment]
My computer doesn't have enough RAM to complete this step. Would it work to just put a previously computed PCA into `self.pcs`?

**Comment by KenjiKamimoto-wustl122** [comment]
Let me answer your question one by one.

I think it is not usual to take long time and large amount of memory in the PCA process.
Even if we use ~50K cells, it took a few minutes and less than 5 GB.
I think you got an unexpected trouble.

Please let me ask one question regarding this problem.
**How many gene do you have in your scRNA-seq dataset?**

If your scRNA-seq data have large number of gene, such as 10K, it might be the cause of your problem. In the tutorial we are recommending to use only highly variable genes. For example, we recommend to pick up top 2000 or 3000 highly variable genes. Please try celloracle after filtering out non-variable genes.

If you still have memory problem after trying the variable gene selection,
I suspect you have another serious problem in the PCA calculation process.
In the celloracle, we use scikit-learn to perform PCA. 
Please go to the scimitar-learn's community to solve your problem.
https://github.com/scikit-learn/scikit-learn
https://scikit-learn.org/stable/

**Comment by KenjiKamimoto-wustl122** [comment]
Please let me answer your second question.

We recommend not changing the code inside celloracle.

Although scanpy performs PCA, the details are not same as the PCA in this process.
If you do that, we never guarantee that celloracle works normally.
You might have unexpected results.
If you still want to do that,  please do so at your own risk and responsibility.

Also, we strongly recommend not skipping PCA and Knn imputation.
It is necessary part.

**Comment by concatenize** [comment]
That makes sense, but I'm not sure how to implement it. It would be obvious how to do it if you were you talking about the [scRNA-seq pre-processing](https://morris-lab.github.io/CellOracle.documentation/notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#6.-Dimensional-reduction). But, we have done that already separately, and I am talking about the [network analysis](https://morris-lab.github.io/CellOracle.documentation/notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data.html#3.1.-PCA). In that notebook, there is a call to the `self.perform_PCA` method, I cannot see how to specify a subset of genes. I could use only a subset of variable genes as input when loading data into the Oracle, but then I would only get regulatory inference involving those genes, right?

**Comment by KenjiKamimoto-wustl122** [comment]
OK, I will add description how to implement the gene selection in the network analysis notebook!

**Comment by KenjiKamimoto-wustl122** [comment]
And, Please let me answer your question below.
>>I could use only a subset of variable genes as input when loading data into the Oracle, but then I would only get regulatory inference involving those genes, right?

It is true, if you only a subset of genes, you will get GRN involving these genes only.
So I guess you might want to use as many genes as possible to infer the GRN.

But we still recommend removing non-variable genes because using non-variable gene does not increase the quality of GRN in most case. GRN inference results might be even worse if you use all genes.
 
I think It is because general scRNA-seq data is noisy and it might be difficult to extract useful information from non-variable gene. 

If you still want to use all genes for GRN inference, we recommend doing GRN analysis with filtered gene first, and then do the same analysis with all genes.

**Comment by concatenize** [comment]
OK, I understand now. Your point it makes sense; it is impossible to infer regulatory relationships from a gene that is roughly constant across all cells (aside from measurement error). But if the gene selection determines input for the regression step, you may want to add that to the preprint. It seems like it could have a big effect on the results depending on whether the filter is strict or more relaxed. This is especially true if the degree of filtering is not pre-specified, because there is often a tendency to optimize results w.r.t. tuning parameters, and the resulting measures of performance can suffer from overfitting.

**Comment by laijen000** [comment]
Hello! 

I have a similar issue. I converted a processed Seurat object into anndata using the command line API. Then for network analysis I loaded the adata and PCA could not complete (because I did not do any gene filtering). I could not find how to implement gene selection in the network analysis notebook -- I would greatly appreciate your advice on how to do this part! Should I take my adata object converted from Seurat and re-do all the pre-processing in scanpy? 

Thank you,
Jenny

**Comment by KenjiKamimoto-wustl122** [comment]
Hi Jenny,

You can do gene selection using scanpy. And you don't need to re-do preprocessing again.
I made example notebook for you. Please use it.
https://github.com/morris-lab/CellOracle/blob/master/docs/notebooks/04_Network_analysis/Network_analysis_with_Seurat_data.ipynb

**Comment by laijen000** [comment]
Thank you for the help and the notebook! 
I was able to carry the analysis to the end of the Network analysis. However, when I tried to get a network score for my gene of interest, there were no scores available. I went back to my adata object and realized that after filtering to keep just the top 3000 variable genes, my TF is not included. My TF is only included if I take the top 15k genes. Is there a way to add my TF back to the filtered adata that contains the top 3000 variable genes? I'm afraid if I use the top 15k the PCA step will run out of memory again. 

Thank you!
